File: bounds.py
#!/usr/bin/env python3
"""Derive the coverage matrix from the inventory and the filesystem.

`corpus.yaml` declares **intent** — for each case, the languages it must
exist in, and any cell deliberately scoped out with a reason. This computes
the **state**: which cells have a fixture, which do not, and the counts.

Nothing is stored. A stored count is a number that can go stale; a derived
one cannot disagree with the tree it describes, and adding a fixture flips
its own cell with no edit to any central file.

    python3 bounds.py            # human summary
    python3 bounds.py --json     # the matrix, for a runner

Exit status is 1 when a declared cell has no fixture, when a fixture exists
that the inventory does not declare, when a case has no `issue_ref`, or when
a positive that names a control has none. Each of those is a way for a
corpus to look complete while measuring less than it claims.
"""
from __future__ import annotations

import json
import pathlib
import re
import sys

import yaml

ROOT = pathlib.Path(__file__).resolve().parent
FIXTURES = ROOT / "fixtures"


class CorpusError(RuntimeError):
    """A declaration and the tree disagree."""


def load_manifest() -> dict:
    return yaml.safe_load((ROOT / "corpus.yaml").read_text())


def _declared(entry) -> tuple[list[str], dict[str, str]]:
    """A cell list is either a bare language list or a scoped-out mapping."""
    if isinstance(entry, list):
        return entry, {}
    languages = entry.get("languages", [])
    out_of_scope = entry.get("out_of_scope", {}) or {}
    for language, reason in out_of_scope.items():
        if not str(reason).strip():
            raise CorpusError(
                f"out_of_scope[{language}] has no reason; an exclusion "
                "without one is a gap wearing a different word"
            )
    return languages, out_of_scope


def discover() -> dict[tuple[str, str, str], pathlib.Path]:
    """Every fixture on disk, keyed by (family, case, language)."""
    found: dict[tuple[str, str, str], pathlib.Path] = {}
    for case_file in sorted(FIXTURES.rglob("case.yaml")):
        rel = case_file.relative_to(FIXTURES).parts
        if len(rel) != 4:
            raise CorpusError(
                f"{case_file} is not at the four-segment path "
                "fixtures/<family>/<case>/<language>/case.yaml"
            )
        family, case, language, _ = rel
        found[(family, case, language)] = case_file.parent
    return found


def producer_criteria(manifest: dict) -> dict[str, dict]:
    """Every criterion each declared producer states, and its reason set.

    Read from the producer's own tree rather than restated here: a restated list
    goes stale the day somebody adds a criterion, and the whole point of the
    number is that nobody can add one without the corpus noticing it is
    unreached.
    """
    out: dict[str, dict] = {}
    for name, declared in (manifest.get("producers") or {}).items():
        pin = ROOT / declared["criteria_pin"]
        pinned = yaml.safe_load(pin.read_text()) if pin.exists() else None
        criteria: set[str] = set(pinned["criteria"]) if pinned else set()
        unreachable: dict[str, str] = {}
        for reason, text in (declared.get("unreachable") or {}).items():
            for identifier in re.findall(
                r"(?:FR|NFR|StR)-\d+-(?:AC|CON|VC)-\d+", text
            ):
                unreachable[identifier] = reason
        out[name] = {
            "root": str(pin),
            "readable": pinned is not None,
            "revision": (pinned or {}).get("revision"),
            "criteria": criteria,
            "unreachable": unreachable,
        }
    return out


def audit() -> dict:
    manifest = load_manifest()
    inventory = manifest["inventory"]
    on_disk = discover()

    cells, gaps, scoped_out, undeclared, problems = [], [], [], [], []

    for family, cases in inventory.items():
        for case, entry in cases.items():
            languages, out_of_scope = _declared(entry)
            for language in languages:
                key = (family, case, language)
                if key in on_disk:
                    cells.append(key)
                else:
                    gaps.append(key)
            for language, reason in out_of_scope.items():
                scoped_out.append((family, case, language, reason))
                if (family, case, language) in on_disk:
                    problems.append(
                        f"{family}/{case}/{language} is scoped out and also "
                        "has a fixture; one of the two is wrong"
                    )

    declared_keys = set(cells) | set(gaps)
    for key in on_disk:
        if key not in declared_keys:
            undeclared.append(key)

    controls = {
        (f, yaml.safe_load((p / "case.yaml").read_text()).get("control_for"), lang)
        for (f, _c, lang), p in on_disk.items()
        if yaml.safe_load((p / "case.yaml").read_text()).get("kind") == "control"
    }
    for (family, case, language), path in sorted(on_disk.items()):
        meta = yaml.safe_load((path / "case.yaml").read_text())
        if not str(meta.get("issue_ref", "")).strip():
            problems.append(f"{family}/{case}/{language} has no issue_ref")
        if "pending" in meta and not str(meta.get("pending", "")).strip():
            problems.append(
                f"{family}/{case}/{language} is pending on nothing; a marker "
                "with no issue is a case nobody comes back to"
            )
        if not (path / "expected.yaml").exists():
            problems.append(f"{family}/{case}/{language} has no expected.yaml")
        if not (path / "input").is_dir():
            problems.append(f"{family}/{case}/{language} has no input tree")
        if meta.get("kind") == "positive" and meta.get("control_for_pair"):
            pair = meta["control_for_pair"]
            # Same language by default. A property that is language independent
            # may point at one control, but the case must name the language it
            # lives in — otherwise "the control is over there somewhere" is
            # indistinguishable from having none.
            where = meta.get("control_language", language)
            if (family, pair, where) not in on_disk:
                problems.append(
                    f"{family}/{case}/{language} names control {pair!r} in "
                    f"{where}, which does not exist"
                )

    # ── criterion coverage ───────────────────────────────────────────────
    claimed: set[str] = set()
    for path in on_disk.values():
        meta = yaml.safe_load((path / "case.yaml").read_text())
        claimed |= set(meta.get("criteria") or [])

    coverage: dict[str, dict] = {}
    for name, declared in producer_criteria(manifest).items():
        if not declared["readable"]:
            coverage[name] = {
                "state": "unavailable",
                "why": f"{declared['root']} is not readable from here",
            }
            continue
        reached = sorted(claimed & declared["criteria"])
        unreachable = {
            identifier: reason
            for identifier, reason in declared["unreachable"].items()
            if identifier in declared["criteria"]
        }
        unreached = sorted(declared["criteria"] - set(reached) - set(unreachable))
        stale = sorted(set(declared["unreachable"]) - declared["criteria"])
        phantom = sorted(claimed - declared["criteria"])
        coverage[name] = {
            "state": "measured",
            "total": len(declared["criteria"]),
            "reached": reached,
            "unreachable": unreachable,
            "unreached": unreached,
            # An id declared unreachable that the producer no longer states, or
            # claimed by a case and stated by nobody. Both are how a coverage
            # number drifts away from the thing it describes.
            "stale_unreachable": stale,
            "claimed_but_undeclared": phantom,
        }
        for identifier in unreached:
            problems.append(
                f"{name} {identifier} is reached by no case and declared "
                "unreachable by nothing"
            )
        for identifier in stale:
            problems.append(
                f"{name} {identifier} is declared unreachable and the producer "
                "no longer states it"
            )
        for identifier in phantom:
            problems.append(
                f"{name} {identifier} is claimed by a case and stated by no "
                "requirement"
            )

    return {
        "criterion_coverage": coverage,
        "covered": sorted(cells),
        "gaps": sorted(gaps),
        "gap_count": len(gaps),
        "scoped_out": sorted(scoped_out),
        "undeclared": sorted(undeclared),
        "problems": sorted(set(problems)),
        "controls": sorted(c for c in controls if c[1]),
    }


def main() -> int:
    try:
        report = audit()
    except CorpusError as err:
        # A structural error is a finding, not a crash: a traceback reads as a
        # broken tool and gets the tool blamed instead of the tree.
        print(f"  PROBLEM    {err}")
        return 1
    if "--json" in sys.argv:
        print(json.dumps(report, indent=1, sort_keys=True))
    else:
        for name, coverage in sorted(report["criterion_coverage"].items()):
            if coverage["state"] != "measured":
                print(f"{name}: criteria unavailable — {coverage['why']}")
                continue
            print(
                f"{name}  {len(coverage['reached'])} reached, "
                f"{len(coverage['unreachable'])} unreachable, "
                f"{len(coverage['unreached'])} unreached  of "
                f"{coverage['total']}"
            )
        print(f"covered      {len(report['covered'])}")
        print(f"gap_count    {report['gap_count']}")
        print(f"scoped out   {len(report['scoped_out'])}")
        for family, case, language in report["gaps"]:
            print(f"  GAP        {family}/{case}/{language}")
        for family, case, language in report["undeclared"]:
            print(f"  UNDECLARED {family}/{case}/{language}")
        for problem in report["problems"]:
            print(f"  PROBLEM    {problem}")
    failed = report["gap_count"] or report["undeclared"] or report["problems"]
    return 1 if failed else 0


if __name__ == "__main__":
    sys.exit(main())

File: src/bounds.rs
//! Corpus inventory discovery and fail-closed bounds reporting.

use std::collections::{BTreeMap, BTreeSet};
use std::path::{Path, PathBuf};

use regex::Regex;
use serde::{Deserialize, Serialize};
use walkdir::WalkDir;

use crate::error::read_text;
use crate::{CorpusError, Result};

/// A stable `(family, case, language)` fixture identity.
#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd, Serialize)]
pub struct CaseKey(pub String, pub String, pub String);

impl CaseKey {
    /// Render the slash-separated public case key.
    #[must_use]
    pub fn slash_name(&self) -> String {
        format!("{}/{}/{}", self.0, self.1, self.2)
    }
}

/// One inventory cell, in either compact or reasoned form.
#[derive(Clone, Debug, Deserialize)]
#[serde(untagged)]
pub enum DeclaredCell {
    /// Every listed language is required.
    Languages(Vec<String>),
    /// Required languages plus explicitly reasoned exclusions.
    Detailed {
        /// Languages this case must still cover.
        #[serde(default)]
        languages: Vec<String>,
        /// Languages this case will never cover, keyed to why.
        #[serde(default)]
        out_of_scope: BTreeMap<String, String>,
    },
}

impl DeclaredCell {
    fn parts(&self) -> Result<(&[String], &BTreeMap<String, String>)> {
        static EMPTY: std::sync::LazyLock<BTreeMap<String, String>> =
            std::sync::LazyLock::new(BTreeMap::new);
        match self {
            Self::Languages(languages) => Ok((languages, &EMPTY)),
            Self::Detailed {
                languages,
                out_of_scope,
            } => {
                for (language, reason) in out_of_scope {
                    if reason.trim().is_empty() {
                        return Err(CorpusError::Invariant(format!(
                            "out_of_scope[{language}] has no reason; an exclusion without one is a gap wearing a different word"
                        )));
                    }
                }
                Ok((languages, out_of_scope))
            }
        }
    }
}

/// A family name held for later work.
#[derive(Clone, Debug, Deserialize)]
pub struct ReservedFamily {
    /// Owning issue.
    pub issue_ref: String,
    /// Why the family is not populated yet.
    pub reason: String,
}

/// One producer's pinned criterion coverage declaration.
#[derive(Clone, Debug, Deserialize)]
pub struct ProducerDeclaration {
    /// Repository-relative pin file.
    pub criteria_pin: String,
    /// Reason name to prose containing criterion identifiers.
    #[serde(default)]
    pub unreachable: BTreeMap<String, String>,
}

/// The parts of `corpus.yaml` interpreted by the bounds engine.
#[derive(Clone, Debug, Deserialize)]
pub struct Manifest {
    /// Declared case/language matrix.
    pub inventory: BTreeMap<String, BTreeMap<String, DeclaredCell>>,
    /// Reserved but deliberately unpopulated families.
    #[serde(default)]
    pub reserved_families: BTreeMap<String, ReservedFamily>,
    /// Producers whose criteria coverage is measured.
    #[serde(default)]
    pub producers: BTreeMap<String, ProducerDeclaration>,
}

/// The case metadata used by inventory and scoring.
#[derive(Clone, Debug, Default, Deserialize)]
pub struct CaseMetadata {
    /// Fixture source language.
    #[serde(default)]
    pub language: String,
    /// Positive/control classification.
    pub kind: Option<String>,
    /// Owning issue.
    pub issue_ref: Option<String>,
    /// An issue that intentionally holds a failing expectation open.
    pub pending: Option<String>,
    /// Positive case named by a control.
    pub control_for: Option<String>,
    /// Control case named by a positive.
    pub control_for_pair: Option<String>,
    /// Non-local control language.
    pub control_language: Option<String>,
    /// Producer criteria reached by this fixture.
    #[serde(default)]
    pub criteria: Vec<String>,
}

/// Derived criterion coverage for one producer.
#[derive(Clone, Debug, Serialize)]
#[serde(tag = "state", rename_all = "kebab-case")]
pub enum CriterionCoverage {
    /// The committed pin could not be read.
    Unavailable {
        /// Human-readable reason the pin file could not be loaded.
        why: String,
    },
    /// Coverage was derived from the pin and fixture claims.
    Measured {
        /// Total criteria declared by the pin.
        total: usize,
        /// Criteria claimed by at least one fixture.
        reached: Vec<String>,
        /// Criteria declared unreachable, keyed to why.
        unreachable: BTreeMap<String, String>,
        /// Criteria reached by no fixture and declared unreachable by nothing.
        unreached: Vec<String>,
        /// Criteria declared unreachable that the pin no longer states.
        stale_unreachable: Vec<String>,
        /// Criteria claimed by a fixture but absent from the pin.
        claimed_but_undeclared: Vec<String>,
    },
}

/// Complete derived bounds report.
#[derive(Clone, Debug, Serialize)]
pub struct BoundsReport {
    /// Per-producer criterion coverage, keyed by producer name.
    pub criterion_coverage: BTreeMap<String, CriterionCoverage>,
    /// Declared cases that exist on disk.
    pub covered: Vec<CaseKey>,
    /// Declared cases missing from disk.
    pub gaps: Vec<CaseKey>,
    /// The number of entries in `gaps`.
    pub gap_count: usize,
    /// `(family, case, language, reason)` tuples excluded from scope.
    pub scoped_out: Vec<(String, String, String, String)>,
    /// Cases found on disk but not declared in the manifest.
    pub undeclared: Vec<CaseKey>,
    /// Human-readable descriptions of every inventory or coverage defect found.
    pub problems: Vec<String>,
    /// Positive cases named as the target of a control case.
    pub controls: Vec<CaseKey>,
}

#[derive(Debug, Deserialize)]
struct CriteriaPin {
    revision: Option<String>,
    #[serde(default)]
    criteria: BTreeSet<String>,
}

/// Load the corpus manifest from `root/corpus.yaml`.
pub fn load_manifest(root: &Path) -> Result<Manifest> {
    let path = root.join("corpus.yaml");
    let text = read_text(&path)?;
    serde_yaml::from_str(&text).map_err(|source| CorpusError::Yaml { path, source })
}

fn load_case_metadata(path: &Path) -> Result<CaseMetadata> {
    let text = read_text(path)?;
    serde_yaml::from_str(&text).map_err(|source| CorpusError::Yaml {
        path: path.to_path_buf(),
        source,
    })
}

/// Discover every `fixtures/<family>/<case>/<language>/case.yaml`.
pub fn discover(root: &Path) -> Result<BTreeMap<CaseKey, PathBuf>> {
    let fixtures = root.join("fixtures");
    let mut case_files = Vec::new();
    for entry in WalkDir::new(&fixtures).follow_links(false) {
        let entry = entry.map_err(|error| {
            CorpusError::Invariant(format!("cannot walk {}: {error}", fixtures.display()))
        })?;
        if entry.file_type().is_file() && entry.file_name() == "case.yaml" {
            case_files.push(entry.into_path());
        }
    }
    case_files.sort();

    let mut found = BTreeMap::new();
    for case_file in case_files {
        let relative = case_file
            .strip_prefix(&fixtures)
            .map_err(|error| CorpusError::Invariant(error.to_string()))?;
        let parts: Vec<_> = relative.iter().collect();
        if parts.len() != 4 {
            return Err(CorpusError::Invariant(format!(
                "{} is not at the four-segment path fixtures/<family>/<case>/<language>/case.yaml",
                case_file.display()
            )));
        }
        let key = CaseKey(
            parts[0].to_string_lossy().into_owned(),
            parts[1].to_string_lossy().into_owned(),
            parts[2].to_string_lossy().into_owned(),
        );
        let parent = case_file.parent().ok_or_else(|| {
            CorpusError::Invariant(format!("{} has no parent", case_file.display()))
        })?;
        found.insert(key, parent.to_path_buf());
    }
    Ok(found)
}

fn parse_unreachable(declaration: &ProducerDeclaration) -> Result<BTreeMap<String, String>> {
    let criterion = Regex::new(r"(?:FR|NFR|StR)-\d+-(?:AC|CON|VC)-\d+")
        .map_err(|error| CorpusError::Invariant(error.to_string()))?;
    let mut out = BTreeMap::new();
    for (reason, text) in &declaration.unreachable {
        for found in criterion.find_iter(text) {
            out.insert(found.as_str().to_owned(), reason.clone());
        }
    }
    Ok(out)
}

/// Derive the complete inventory, control, and criterion-coverage report.
pub fn audit(root: &Path) -> Result<BoundsReport> {
    let manifest = load_manifest(root)?;
    let on_disk = discover(root)?;
    let mut cells = BTreeSet::new();
    let mut gaps = BTreeSet::new();
    let mut scoped_out = BTreeSet::new();
    let mut problems = BTreeSet::new();

    for (family, cases) in &manifest.inventory {
        for (case, declaration) in cases {
            let (languages, exclusions) = declaration.parts()?;
            for language in languages {
                let key = CaseKey(family.clone(), case.clone(), language.clone());
                if on_disk.contains_key(&key) {
                    cells.insert(key);
                } else {
                    gaps.insert(key);
                }
            }
            for (language, reason) in exclusions {
                scoped_out.insert((
                    family.clone(),
                    case.clone(),
                    language.clone(),
                    reason.clone(),
                ));
                let key = CaseKey(family.clone(), case.clone(), language.clone());
                if on_disk.contains_key(&key) {
                    problems.insert(format!(
                        "{}/{}/{} is scoped out and also has a fixture; one of the two is wrong",
                        family, case, language
                    ));
                }
            }
        }
    }

    let declared: BTreeSet<_> = cells.union(&gaps).cloned().collect();
    let undeclared: BTreeSet<_> = on_disk
        .keys()
        .filter(|key| !declared.contains(*key))
        .cloned()
        .collect();
    let mut controls = BTreeSet::new();
    let mut claimed = BTreeSet::new();

    for (key, path) in &on_disk {
        let metadata = load_case_metadata(&path.join("case.yaml"))?;
        claimed.extend(metadata.criteria);
        if metadata.kind.as_deref() == Some("control")
            && let Some(control_for) = metadata.control_for.filter(|value| !value.is_empty())
        {
            controls.insert(CaseKey(key.0.clone(), control_for, key.2.clone()));
        }
        if metadata
            .issue_ref
            .as_deref()
            .is_none_or(|value| value.trim().is_empty())
        {
            problems.insert(format!("{} has no issue_ref", key.slash_name()));
        }
        if metadata
            .pending
            .as_deref()
            .is_some_and(|value| value.trim().is_empty())
        {
            problems.insert(format!(
                "{} is pending on nothing; a marker with no issue is a case nobody comes back to",
                key.slash_name()
            ));
        }
        if !path.join("expected.yaml").is_file() {
            problems.insert(format!("{} has no expected.yaml", key.slash_name()));
        }
        if !path.join("input").is_dir() {
            problems.insert(format!("{} has no input tree", key.slash_name()));
        }
        if metadata.kind.as_deref() == Some("positive")
            && let Some(pair) = metadata.control_for_pair
        {
            let language = metadata.control_language.unwrap_or_else(|| key.2.clone());
            if !on_disk.contains_key(&CaseKey(key.0.clone(), pair.clone(), language.clone())) {
                problems.insert(format!(
                    "{} names control {pair:?} in {language}, which does not exist",
                    key.slash_name()
                ));
            }
        }
    }

    let mut criterion_coverage = BTreeMap::new();
    for (name, declaration) in &manifest.producers {
        let pin_path = root.join(&declaration.criteria_pin);
        let Ok(pin_text) = std::fs::read_to_string(&pin_path) else {
            criterion_coverage.insert(
                name.clone(),
                CriterionCoverage::Unavailable {
                    why: format!("{} is not readable from here", pin_path.display()),
                },
            );
            continue;
        };
        let pin: CriteriaPin =
            serde_yaml::from_str(&pin_text).map_err(|source| CorpusError::Yaml {
                path: pin_path.clone(),
                source,
            })?;
        let _revision = pin.revision;
        let unreachable = parse_unreachable(declaration)?;
        let reached: Vec<_> = claimed.intersection(&pin.criteria).cloned().collect();
        let unreached: Vec<_> = pin
            .criteria
            .difference(&claimed)
            .filter(|identifier| !unreachable.contains_key(*identifier))
            .cloned()
            .collect();
        let stale: Vec<_> = unreachable
            .keys()
            .filter(|identifier| !pin.criteria.contains(*identifier))
            .cloned()
            .collect();
        let phantom: Vec<_> = claimed.difference(&pin.criteria).cloned().collect();
        for identifier in &unreached {
            problems.insert(format!(
                "{name} {identifier} is reached by no case and declared unreachable by nothing"
            ));
        }
        for identifier in &stale {
            problems.insert(format!(
                "{name} {identifier} is declared unreachable and the producer no longer states it"
            ));
        }
        for identifier in &phantom {
            problems.insert(format!(
                "{name} {identifier} is claimed by a case and stated by no requirement"
            ));
        }
        criterion_coverage.insert(
            name.clone(),
            CriterionCoverage::Measured {
                total: pin.criteria.len(),
                reached,
                unreachable,
                unreached,
                stale_unreachable: stale,
                claimed_but_undeclared: phantom,
            },
        );
    }

    Ok(BoundsReport {
        criterion_coverage,
        covered: cells.into_iter().collect(),
        gaps: gaps.iter().cloned().collect(),
        gap_count: gaps.len(),
        scoped_out: scoped_out.into_iter().collect(),
        undeclared: undeclared.into_iter().collect(),
        problems: problems.into_iter().collect(),
        controls: controls.into_iter().collect(),
    })
}

impl BoundsReport {
    /// Whether any fail-closed inventory condition is present.
    #[must_use]
    pub fn failed(&self) -> bool {
        self.gap_count != 0 || !self.undeclared.is_empty() || !self.problems.is_empty()
    }
}
